ethclient/httpsub.go

```diff
@@ -0,0 +1,150 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package ethclient
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// subscribeNewHeadPolling creates a block filter and polls it.
+// This is the fallback for SubscribeNewHead on HTTP connections.
+func (ec *Client) subscribeNewHeadPolling(ctx context.Context, ch chan<- *types.Header) (event.Subscription, error) {
+	var id string
+	err := ec.c.CallContext(ctx, &id, "eth_newBlockFilter")
+	if err != nil {
+		return nil, err
+	}
+	sub := newFilterSub(ec.c, id, ch)
+	return sub, nil
+}
+
+// subscribeFilterLogs subscribes to a log filter.
+// This is the fallback for SubscribeFilterLogs on HTTP connections.
+func (ec *Client) subscribeFilterLogsPolling(ctx context.Context, q ethereum.FilterQuery, ch chan<- types.Log) (ethereum.Subscription, error) {
+	arg, err := toFilterArg(q)
+	if err != nil {
+		return nil, err
+	}
+	var id string
+	err = ec.c.CallContext(ctx, &id, "eth_newFilter", arg)
+	if err != nil {
+		return nil, err
+	}
+	sub := newFilterSub(ec.c, id, ch)
+	return sub, nil
+}
+
+// filterSub implements event.Subscription with the polling filter API.
+type filterSub[Result any] struct {
+	// The channel to send the results to
+	resultc   chan<- Result
+	errc      chan error
+	closed    chan struct{}
+	unsubOnce sync.Once
+
+	id     string
+	client *rpc.Client
+}
+
+func newFilterSub[Result any](client *rpc.Client, id string, resultc chan<- Result) *filterSub[Result] {
+	sub := &filterSub[Result]{
+		resultc: resultc,
+		errc:    make(chan error, 1),
+		closed:  make(chan struct{}),
+		client:  client,
+	}
+	go sub.poll()
+	return sub
+}
+
+// Unsubscribe cancels the event subscription.
+func (s *filterSub[Result]) Unsubscribe() {
+	s.closeWithError(nil)
+}
+
+// Err returns the subscription error channel.
+func (s *filterSub[Result]) Err() <-chan error {
+	return s.errc
+}
+
+func (s *filterSub[Result]) poll() {
+	var timer = time.NewTicker(10 * time.Second)
+	for {
+		err := s.getChanges()
+		if err != nil {
+			s.closeWithError(err)
+			return
+		}
+
+		// Wait for next time.
+		select {
+		case <-timer.C:
+		case <-s.closed:
+			return
+		}
+	}
+}
+
+func (s *filterSub[Result]) closeWithError(err error) {
+	s.unsubOnce.Do(func() {
+		close(s.closed)
+		unsubErr := s.uninstallFilter()
+		if unsubErr != nil && err == nil {
+			err = unsubErr
+		}
+		if err != nil {
+			select {
+			case s.errc <- err:
+			default:
+			}
+		}
+		close(s.errc)
+	})
+}
+
+// getChanges calls eth_getFilterChanges and delivers the results.
+func (s *filterSub[Result]) getChanges() error {
+	ctx, done := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
should the 10 seconds be pulled out into a constant and shared with the poller?

Reply:
I'm not super sure about this fixed 10s delay.

Reply:
Is there a reason to even have a timeout here? After all, there is no better thing to do than wait if the query gets stuck. The context could just be cancelled on `closeWithError` (which would be a nice thing to do anyways even if there is a timeout).